### Hospital Readmission Prediction Using Machine Learning on the Diabetes 130-US Hospitals Dataset (1999–2008)

---

### Dataset

Diabetes 130-US Hospitals for Years 1999–2008

---

### Project Overview

Hospital readmission is a significant challenge in modern healthcare systems because frequent readmissions increase healthcare costs, place additional pressure on hospitals and may indicate that a patient's condition was not effectively managed. Predicting whether a patient is likely to be readmitted allows healthcare providers to identify high-risk patients, improve treatment planning and allocate medical resources more efficiently.

This project investigates the **Diabetes 130-US Hospitals for Years 1999–2008** dataset obtained from the **UCI Machine Learning Repository**. The objective is to develop and evaluate multiple supervised machine learning models capable of predicting patient readmission based on demographic information, hospital admission details, laboratory results, diagnoses and prescribed diabetic medications.

Throughout this notebook we will demonstrate a systematic machine learning methodology.

---

### Notebook Objectives

The primary objectives of this notebook are to:

- Understand the characteristics and quality of the dataset.
- Perform comprehensive exploratory data analysis (EDA).
- Identify and resolve data quality issues through appropriate preprocessing techniques.
- Engineer and select informative features for predictive modelling.
- Train and compare multiple supervised machine learning classification algorithms.
- Optimise model performance using appropriate hyperparameter tuning techniques.
- Evaluate model performance using multiple classification metrics.
- Identify the most suitable model for predicting hospital readmission while discussing performance, interpretability and practical applicability.

---

### Notebook Workflow

The project follows the standard machine learning workflow illustrated below.

```

Dataset Collection
↓
Exploratory Data Analysis (EDA)
↓
Data Preprocessing
↓
Feature Engineering
↓
Feature Selection
↓
Train-Test Split
↓
Model Development
↓
Hyperparameter Optimisation
↓
Model Evaluation
↓
Model Comparison
↓
Conclusion

```

Each stage of this workflow is presented with the underlying theory, implementation, experimental results and critical discussion to provide a complete understanding of the machine learning pipeline.

---

## 1. Dataset Overview

Before building any machine learning model, it is important to understand the dataset. A machine learning model learns patterns from the data, so the quality of the data has a direct impact on the quality of the predictions. If the dataset contains missing values, incorrect information or unnecessary features, the model may produce poor results.

In this section, we will introduce the dataset, explain why it was chosen, describe what information it contains and discuss how it was collected. Understanding the dataset first will help us make better decisions later during data preprocessing, feature engineering and model building.

---

### 1.1 Introduction to the Dataset

The dataset used in this project is the **Diabetes 130-US Hospitals for Years 1999–2008** dataset obtained from the **UCI Machine Learning Repository**.

The dataset contains **101,766 inpatient hospital encounters** collected from **130 hospitals and integrated healthcare delivery networks** across the United States between **1999 and 2008**.

Each row in the dataset represents **one hospital encounter**, not one patient. This means that the same patient may appear more than once if they were admitted to the hospital multiple times during the ten-year period. Each column contains information describing that hospital encounter, such as the patient's age, diagnosis, medications, laboratory test results and previous hospital visits.

The goal of this project is to use this information to predict whether a patient will be readmitted to the hospital after discharge.

---

### 1.2 Why was this dataset collected?

To make sure the data represented meaningful hospital admissions, only encounters that satisfied the following conditions were included:

1. The encounter was an **inpatient hospital admission**.
2. The patient had **diabetes recorded as one of the diagnoses**.
3. The patient stayed in the hospital for **at least one day but no more than fourteen days**.
4. **Laboratory test** was performed during the hospital stay.
5. **Medication** was given during the hospital encounter.

These conditions ensure that every record contains enough clinical information to support meaningful analysis and machine learning.

---

### 1.3 Why was this dataset selected?

The Diabetes 130-US Hospitals dataset was chosen for several reasons.

- It addresses a real-world healthcare problem with practical importance.
- It contains **101,766 hospital encounters**, providing enough data to train and evaluate machine learning models.
- It includes information from **130 hospitals** collected over a **10-year period (1999–2008)**, making it a large and diverse healthcare dataset.
- It contains both numerical and categorical features, allowing different preprocessing and encoding techniques to be demonstrated.
- It includes real-world data challenges such as missing values, high-cardinality categorical variables and class imbalance, making it suitable for demonstrating data preprocessing and feature engineering.
- It contains demographic, clinical, laboratory and medication information, allowing many different factors that may influence hospital readmission to be analysed.
- It is a well-known public dataset that has been widely used in healthcare and machine learning research, making it suitable for comparison with previous studies.

- Overall, this dataset provides an excellent opportunity to demonstrate the complete machine learning workflow, from understanding the data and preparing it for analysis to building, optimising and evaluating multiple machine learning models.
---

### 1.4 What information does the dataset contain?

The dataset contains more than fifty features describing different aspects of a patient's hospital encounter. Instead of looking at every feature individually, it is easier to group similar features together.

### Demographic Information

These features describe the patient.

Examples include:

- Race
- Gender
- Age
- Weight

These features provide background information about each patient and may influence hospital readmission.

### Hospital Admission Information

These features describe the patient's hospital stay.

Examples include:

- Admission type
- Admission source
- Discharge disposition
- Time spent in hospital

These features describe how the patient entered the hospital, how long they stayed and how they were discharged.

### Clinical Information

These features describe the patient's medical condition.

Examples include:

- Primary diagnosis
- Secondary diagnoses
- Number of diagnoses
- Medical specialty

These features help describe the patient's overall health and medical history.

### Laboratory Information

These features describe laboratory tests performed during the hospital stay.

Examples include:

- HbA1c test result
- Serum glucose result
- Number of laboratory procedures

Laboratory test results provide useful information about the patient's condition during admission.

### Medication Information

The dataset also contains several features describing diabetes medications.

These features record whether a medication was prescribed and whether the dosage was increased, decreased or kept the same during the hospital stay.

Medication information may provide useful clues about the patient's treatment and disease management.

### Previous Healthcare Utilisation

These features describe how often the patient used healthcare services before the current hospital admission.

Examples include:

- Number of outpatient visits
- Number of emergency visits
- Number of inpatient visits

 These features describe the patient's use of healthcare services during the year before the current hospital admission. A patient's previous healthcare utilisation may provide useful information about their health status and may help machine learning models predict the risk of future hospital readmission.
 
---

### 1.5 What will we do next?

Now that we understand the background of the dataset, the next step is to load it into Python and explore its structure.

We will examine:

- the number of rows and columns,
- the data types of each feature,
- missing values,
- summary statistics,
- and the target variable.

This will help us understand the quality of the data and decide what preprocessing steps are required before building machine learning models.

---

### Load csv from file

Read a CSV file from your local system. 

The `.read_csv()` function returns a `pandas.DataFrame` object, which is a powerful 2D data structure that allows row and column operations, descriptive statistics, and data manipulations. 

---


In [ ]:
# Install required libraries
%pip install pandas


In [ ]:
import pandas as pd

# ============================================================
# Load the Diabetes Dataset
# ============================================================
filename = "diabetic_data.csv"

# Read the dataset while preserving valid category values
# such as "None". This prevents pandas from automatically
# converting them into missing values (NaN).

df = pd.read_csv(
    filename,
    keep_default_na=False
)

### Checking the Dataset

*After loading a dataset, it is good practice to inspect the first few rows before performing any analysis. This helps confirm that the dataset has been loaded correctly and provides a quick overview of its structure.*

*The `head()` function displays the first five rows of a DataFrame by default. It is commonly used to verify that the column names have been read correctly and that the values appear under the correct features.*

We will now display the first five rows of the dataset to confirm that the data has been loaded successfully.

---

In [32]:
# ============================================================
# Display the First Five Rows
# ============================================================

df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


### Understanding the Output

The output displays the first five hospital encounters in the dataset.

From these rows, we can make a few initial observations.

- Each row represents one hospital encounter.
- Each column represents one feature describing that encounter.
- The dataset contains a mixture of numerical and categorical features.
- The column names appear to have been loaded correctly.
- The dataset appears to have been imported successfully.

At this stage, we are only confirming that the dataset has been loaded correctly. A detailed exploration of the data will be carried out in the upcoming sections.

---

### 1.6 Exploring the Dataset

### What is the dimensionality of our data?

Before we start analysing or training a model, it's helpful to know the *shape* of the dataset, that is, how many rows and columns it contains. This gives us a sense of how big the dataset is and what we're working with.

We can check this using the `.shape` attribute in a pandas DataFrame, which returns a result like this:

```
(rows, columns)
```

- The number of *rows* tells us how many *examples* or *participants* are in the dataset.  
- The number of *columns* tells us how many *features* (measurements) we have for each example. 

Example:

- Each **row** represents one hospital encounter.
- Each **column** represents one feature describing that encounter.

Understanding the dataset's shape helps us:
- Plan how to split the data into training and testing sets  
- Estimate how much memory we'll need  
- Choose models and techniques that are suitable for the size of the data  


In [33]:
# How big is our data
print(df.shape)

(101766, 50)



This knowledge helps us ensure we have the complete dataset loaded. It's also a good initial check before we proceed to more in-depth data profiling or feature analysis.

The output is:

```
(101766, 50)
```

It means:

- The dataset contains **101,766 hospital encounters**.
- Each encounter is described using **50 features**.

---

### Exploring more details of the data

Beyond checking the first few and last few rows, pandas offers convenient methods to quickly summarise your dataset. For instance, if you want to view column data types and any non-null counts, you can use `data.info()`. This can be useful to spot missing values or confirm that columns are numeric:

In [34]:
# Checking data info
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   encounter_id              101766 non-null  int64
 1   patient_nbr               101766 non-null  int64
 2   race                      101766 non-null  str  
 3   gender                    101766 non-null  str  
 4   age                       101766 non-null  str  
 5   weight                    101766 non-null  str  
 6   admission_type_id         101766 non-null  int64
 7   discharge_disposition_id  101766 non-null  int64
 8   admission_source_id       101766 non-null  int64
 9   time_in_hospital          101766 non-null  int64
 10  payer_code                101766 non-null  str  
 11  medical_specialty         101766 non-null  str  
 12  num_lab_procedures        101766 non-null  int64
 13  num_procedures            101766 non-null  int64
 14  num_medications           10176

From `data.info()`, we get a quick overview of the dataset's structure. It tells us:

- How many rows are in the dataset  
- Which columns are present  
- How many non-null (non-missing) values are in each column  
- What data type each column uses (such as `int64` for whole numbers, `float64` for decimal numbers, or `object` for text)

This information is especially helpful for spotting issues early. For example:
- If a column has fewer non-null values than the total number of rows, it likely contains *missing data*.
- If a column that should contain numbers is listed as `object`, the values might be stored as text, possibly due to formatting issues (like commas or symbols).

Checking the results of `data.info()`, allows us to make sure the dataset is properly formatted and ready for analysis or model training.

---

### 1.7 Analysing , visualising and Transforming tabular data

### Introduction  

We will explore how to analyse and preprocess a numeric/tabular dataset to gain meaningful insights before applying machine learning models. Understanding your data is a crucial first step in any data science workflow, as it influences how well your models perform and generalise to unseen data.

We will cover three key aspects of data exploration and preparation:  

- *Describing the dataset*: we will generate statistical summaries, such as measures of central tendency (mean, median) and dispersion (variance, standard deviation), to understand the distribution of each feature. Additionally, we will examine missing values, outliers, and correlations between variables to identify patterns and potential preprocessing steps.  

- *Visualising the dataset*: data visualisation helps us uncover hidden relationships, trends, and anomalies. We will use histograms, box plots, scatter plots, and correlation heatmaps to visually inspect distributions, detect skewness, and understand feature interactions.    
**Visualisations will be introduced throughout the notebook whenever they help explain the data. This approach ensures that each visualisation has a clear purpose rather than being included unnecessarily.**

This process is essential for making informed decisions in machine learning, as the quality of your input data directly impacts model accuracy and reliability.

### Describing the Dataset

#### Statistical summary

Descriptive statistics help us understand how each part of a dataset is spread out. When we run `.describe()` on a pandas DataFrame in Python, we get a helpful summary that includes:

- *Count*:  </br>
Tells us how many values are present (not missing) for each column. This helps us spot any gaps in the data.

- *Mean*:  </br>
The average value. This gives us a rough idea of what's "typical" for each column.

- *Standard Deviation*:</br>
Shows how much the values vary. A high number means the data is more spread out.

- *Minimum Value*:  </br>
The smallest number in the column. This can help us find any errors or outliers.

- *25th Percentile (Q1)*:  </br>
A quarter of the values fall below this point. It gives us a sense of the lower end of the data.

- *50th Percentile (Median/Q2)*:  </br>
The middle value when the data is sorted. This shows us the centre of the data.

- *75th Percentile (Q3)*:  </br>
Three-quarters of the data fall below this point. It helps us understand the upper range.

- *Maximum Value*:  </br>
The largest value in the column. This can highlight extreme values or mistakes.

These summaries make it easier to spot missing information, unusual results, or parts of the data that might need special attention, like features with a very wide range of values:


In [35]:
# Summary Statistics 
df.describe()

,encounter_id,patient_nbr,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,number_diagnoses
count,1.017660e+05,1.017660e+05,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000
mean,1.652016e+08,5.433040e+07,2.024006,3.715642,5.754437,4.395987,43.095641,1.339730,16.021844,0.369357,0.197836,0.635566,7.422607
std,1.026403e+08,3.869636e+07,1.445403,5.280166,4.064081,2.985108,19.674362,1.705807,8.127566,1.267265,0.930472,1.262863,1.933600
min,1.252200e+04,1.350000e+02,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,8.496119e+07,2.341322e+07,1.000000,1.000000,1.000000,2.000000,31.000000,0.000000,10.000000,0.000000,0.000000,0.000000,6.000000
50%,1.523890e+08,4.550514e+07,1.000000,1.000000,7.000000,4.000000,44.000000,1.000000,15.000000,0.000000,0.000000,0.000000,8.000000
75%,2.302709e+08,8.754595e+07,3.000000,4.000000,7.000000,6.000000,57.000000,2.000000,20.000000,0.000000,0.000000,1.000000,9.000000
max,4.438672e+08,1.895026e+08,8.000000,28.000000,25.000000,14.000000,132.000000,6.000000,81.000000,42.000000,76.000000,21.000000,16.000000


To explore the categorical features in the Diabetes Hospital Readmission dataset, we use the `describe()` function with the parameter `include=["object"]`. This tells pandas to generate summary statistics only for columns stored as the **object** data type, which in this dataset represents the categorical features.

The output of `describe()` is displayed using **`.T` (transpose)**, which swaps the rows and columns of the table. This does not change the data itself; it simply makes the output easier to read by displaying one feature per row.


In [36]:
# Summary Statistics for Categorical Features
df.describe(include=["object"]).T

/var/folders/6w/6jhfgq8s7hb_9__z7h3g3tcm0000gn/T/ipykernel_14159/4028559741.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include=["object"]).T


,count,unique,top,freq
race,101766,6,Caucasian,76099
gender,101766,3,Female,54708
age,101766,10,[70-80),26068
weight,101766,10,?,98569
payer_code,101766,18,?,40256
medical_specialty,101766,73,?,49949
diag_1,101766,717,428,6862
diag_2,101766,749,276,6752
diag_3,101766,790,250,11555
max_glu_serum,5346,3,Norm,2597


### Understanding the Output

Unlike numerical features, categorical features are summarised using different statistics because they represent labels or categories rather than continuous numerical values.

The output includes:

- **Count** – The number of values that are **not recognised by pandas as missing**.
- **Unique** – The number of distinct categories present in the feature.
- **Top** – The most frequently occurring category (mode).
- **Freq** – The number of times the most common category appears.

These statistics provide a quick overview of the categorical features and help us understand the distribution of different categories before performing data preprocessing or feature engineering.

An important point to note is that these statistics may not always represent the true characteristics of the data. In this dataset, some missing information is stored using placeholder values such as `?` . Since pandas treats `?` as ordinary text, these placeholder values are included in the calculated statistics. As a result, the **Count**, **Unique**, **Top** and **Freq** values may be affected.

For this reason, a dedicated missing value analysis will be performed later in the notebook.

---

## 2. Data Quality Assessment

*Before building any machine learning model, it is essential to assess the quality of the dataset. A machine learning model learns patterns directly from the data it is given. Therefore, if the dataset contains missing values, incorrect data, duplicate records or inconsistent formatting, these problems can negatively affect the model's performance and reduce its ability to make accurate predictions on unseen data.*

*Data Quality Assessment is the process of examining the dataset to identify issues that may need to be addressed before further analysis or model training. Detecting these problems early allows appropriate preprocessing techniques to be applied and helps ensure that the final dataset is suitable for machine learning.*

*The main aspects of data quality assessment include:*

- *Identifying missing values.*
- *Detecting duplicate records.*
- *Checking data types and feature consistency.*
- *Identifying placeholder values.*
- *Assessing whether any features contain too little useful information.*

For the Diabetes Hospital Readmission dataset, this stage is particularly important because several features are known to contain missing information represented by placeholder values such as `?`. As a result, these issues are not immediately detected by functions such as `info()` or `describe()` and require further investigation.

---

### What will we examine?

To assess the quality of the dataset, we will investigate several important aspects of the data.

1. **Missing Value Analysis**  
   Identify features containing missing information and determine how much data is missing.

2. **Duplicate Record Analysis**  
   Check whether duplicate hospital encounters exist within the dataset.

3. **Data Type Verification**  
   Verify that each feature has an appropriate data type and identify any features that may require conversion during preprocessing.

4. **Feature Quality Assessment**  
   Examine individual features to identify those that contain very little useful information, a large number of unique categories or other characteristics that may influence preprocessing decisions.

The outcome of this assessment will determine how the dataset is cleaned and prepared before feature engineering, feature selection and machine learning model development.

---

### Missing Value Analysis

*Missing values are one of the most common data quality problems encountered in real-world datasets. They can reduce the amount of information available to a machine learning model and may negatively affect prediction performance if they are not handled appropriately.*

*It is important to identify which features contain missing information and determine the extent of the missing data. This helps us make informed preprocessing decisions rather than making assumptions about the quality of the dataset.*

*In many real-world datasets, placeholder values such as `?` are used as an actual missing value. Since pandas treats these placeholders as ordinary text, they must first be converted into proper missing values before further analysis can be performed.*

In this section, we will first convert placeholder values into missing values recognised by pandas. We will then calculate the number and percentage of missing values for each feature to identify columns that may require further preprocessing.

---

In [ ]:
# ============================================================
# Convert Placeholder Values to Missing Values
# ============================================================

# The Diabetes dataset uses '?' to represent missing information.
# Pandas treats '?' as ordinary text rather than a missing value.
# Replace every '?' with pd.NA so that pandas recognises it as missing.

df.replace("?", pd.NA, inplace=True)

### Calculating Missing Values

*After converting placeholder values into missing values recognised by pandas, the next step is to measure the extent of missing data within the dataset.*

*Missing value analysis helps identify which features contain incomplete information and how much data is missing from each feature. This information is important because different levels of missing data may require different preprocessing strategies.*

*Both the number of missing values and the percentage of missing values are calculated. To show a better indication of the overall severity of missing data relative to the size of the dataset.*

The results will be combined into a single summary table and sorted from the highest percentage of missing values to the lowest. This makes it easier to identify the features that require the greatest attention during data preprocessing.

---

In [38]:
# ============================================================
# Count Missing Values
# ============================================================

# Count the number of missing values in each feature.
# isnull() identifies missing values (NaN / pd.NA).
# sum() adds the True values to calculate the total number
# of missing values for every column.

missing_count = df.isnull().sum()

# ============================================================
# Calculate Missing Value Percentage
# ============================================================

# Calculate the percentage of missing values for each feature.
# Divide the missing value count by the total number of rows
# and multiply by 100.

missing_percentage = ((missing_count / len(df)) * 100).round(2)

# ============================================================
# Create Missing Value Summary Table
# ============================================================

# Create a summary table containing both the number and
# percentage of missing values for each feature.

missing_summary = pd.DataFrame({

    "Missing Count": missing_count,

    "Missing Percentage (%)":
        missing_percentage

})

# Sort the table so that the features with the highest
# percentage of missing values appear first.

missing_summary = missing_summary.sort_values(
    by="Missing Percentage (%)",
    ascending=False
)

# Display the missing value summary.

missing_summary

,Missing Count,Missing Percentage (%)
weight,98569,96.86
max_glu_serum,96420,94.75
A1Cresult,84748,83.28
medical_specialty,49949,49.08
payer_code,40256,39.56
race,2273,2.23
diag_3,1423,1.40
diag_2,358,0.35
diag_1,21,0.02
encounter_id,0,0.00


### Observations

The missing value summary shows that most features in the dataset are complete and contain no missing values. Only a small number of features contain missing information, although the amount of missing data varies considerably between them.

At this stage, no preprocessing decisions are made. The missing value analysis is used to understand the quality of the dataset and will guide the preprocessing decisions in the following sections.

### 2.2 Duplicate Record Analysis

*Duplicate records occur when exactly the same observation appears more than once in a dataset. Identifying duplicate records is an important part of data quality assessment because duplicated observations can introduce bias into statistical analysis and machine learning models.*

*In this section, the dataset will be examined to determine whether any completely identical rows exist. If duplicate records are found, they will be investigated before deciding whether they should be removed.*

---

In [ ]:
# ============================================================
# Check for Duplicate Records
# ============================================================

# Count the number of duplicate rows in the dataset.
# duplicated() identifies rows that are exact copies of
# previous rows.
# sum() counts the total number of duplicate records.

duplicate_count = df.duplicated().sum()

print(f"Number of duplicate records: {duplicate_count}")

### Observations

No duplicate records were identified in the dataset. This indicates that each hospital encounter is unique and no duplicate rows require removal before further preprocessing.

It is important to note that this does not mean every patient appears only once. Since each row represents a hospital encounter, the same patient may appear multiple times if they were admitted to hospital on different occasions.

---

### 2.3 Initial Feature Assessment

*Not every feature contained within a dataset is necessarily useful for machine learning. Some variables exist only to uniquely identify observations, while others may contain little predictive information or present significant data quality issues.*

*Before performing data preprocessing, it is useful to examine the purpose of some features . Removing unnecessary features reduces model complexity, decreases computational cost and prevents algorithms from learning patterns from information that is unrelated to the prediction task.*

*At this stage, features are assessed based on their meaning, their role within the dataset and the results of the previous data quality assessment. Final preprocessing decisions will then be made before preparing the dataset for machine learning.*

Feature: `encounter_id`

Unique identifier of  each hospital encounter within the dataset. It serves only as an  identifier and does not describe the patient's health condition, treatment or likelihood of hospital readmission, it is unlikely to provide useful predictive information for machine learning.Therefore, this feature is considered a candidate for removal during preprocessing.


Feature: `patient_nbr`

Uniquely identifies individual patients. Same patient may appear multiple times because each row represents a separate hospital encounter, this feature represents patient identity rather than clinical characteristics. Furthermore, the dataset already contains features such as `number_outpatient`, `number_emergency` and `number_inpatient`, which summarise the patient's previous healthcare utilisation without revealing their identity. Therefore, this feature is also considered a candidate for removal before model training.


Feature: `weight`

The `weight` feature contains 96.86% missing values, making it the feature with the largest amount of missing information in the dataset. Such a high proportion of missing data leaves very little usable information for model training and would make reliable imputation difficult. Therefore, `weight` is considered a strong candidate for removal during preprocessing.


Feature: `payer_code`

The `payer_code` feature contains approximately 39.56% missing values. Although this is considerably lower than the `weight` feature, a substantial proportion of the information is still unavailable. As this variable primarily records Integer identifier corresponding to 23 distinct values, for example, Blue Cross/Blue Shield, Medicare, and self-pay. Therefore, this feature is identified for further investigation before making a preprocessing decision.


Feature: `medical_specialty`

The `medical_specialty` feature contains approximately 49.08% missing values. Despite this, it may still provide clinically useful information because it records Integer identifier of a specialty of the admitting physician. This feature will also be investigated further before deciding how it should be handled during preprocessing.

### Summary

The initial feature assessment focused only on the features that required immediate attention because they either act as identifiers or contain a substantial proportion of missing values.

The remaining features were not included in this discussion, this does not necessarily mean they will all be retained without modification. Many of these features, particularly the categorical variables and medication-related attributes, will require further preprocessing, including missing value treatment, category encoding and data transformation.

Therefore, the assessment presented in this section should be considered an initial evaluation rather than a complete review of every feature. The remaining features will be examined individually during the subsequent preprocessing stages as required.

---

## 3. Data Preprocessing

*The previous section focused on understanding the quality and characteristics of the dataset. Based on these findings, the dataset can now be prepared for machine learning.*

*Data preprocessing involves modifying the dataset so that it is suitable for training machine learning models. This includes removing unnecessary features, handling missing values, transforming categorical variables and preparing numerical features for analysis.*

*The preprocessing steps are performed in a logical sequence so that each decision is supported by the results of the previous data quality assessment.*

---

### 3.1 Feature Removal

*The initial feature assessment identified features that are unlikely to contribute useful information for predicting hospital readmission. Removing these features before performing other preprocessing tasks simplifies the dataset and avoids unnecessary processing of variables that will not be used during model training.*

*Each removal decision is based on the purpose of the feature and the observations made during the data quality assessment.*

`Features Selected for Removal`

Based on the data quality assessment and the initial feature assessment, three features have been identified for removal before further preprocessing.

`encounter_id`
Since it provides no meaningful predictive information, this feature will be removed.

`patient_nbr`
Previous healthcare utilisation is already represented by other variables in the dataset, making this identifier unnecessary for prediction. Therefore, this feature will be removed.

`weight`
Contains approximately 96.86% missing values, meaning that weight information is unavailable for most hospital encounters. With such a large proportion of missing data, the feature provides very little useful information for model training and reliable imputation would be difficult. Therefore, this feature will be removed.

Features such as `payer_code` and `medical_specialty` require further investigation because they contain useful information despite having missing values. 

Other categorical and numerical features will also undergo additional preprocessing before they are used for machine learning.

In [ ]:
# ============================================================
# Remove Unnecessary Features
# ============================================================

# Remove features that were identified as unsuitable
# for machine learning during the feature assessment.

df = df.drop(
    columns=[
        "encounter_id",
        "patient_nbr",
        "weight"
    ]
)

# Display the updated dataset dimensions.

print("Updated dataset shape:", df.shape)

### 3.2 Missing Value Treatment

*Following the feature removal, remaining missing values must be addressed before the dataset can be used for machine learning.*

*Missing values should not all be handled using the same technique. The most appropriate treatment depends on the proportion of missing values, the purpose of the feature and the potential impact on model performance.*

*In this section, each remaining feature containing missing values will be examined individually. Based on its characteristics, an appropriate preprocessing strategy will be selected before any modifications are applied to the dataset.*

Feature: `payer_code`

The `payer_code` feature records the patient's payment or insurance category for the hospital encounter. It does not directly describe the patient's clinical condition.

Approximately **39.56%** of the values are missing, indicating that a substantial proportion of the information is unavailable. Retaining this feature would require replacing a large number of missing values which is likely to be less relevant than clinical variables such as diagnoses, laboratory results and medications.

For these reasons, the `payer_code` feature will be removed before model training.

### Features Retained for Further Preprocessing

The following features were retained because they contain potentially useful clinical or demographic information for predicting hospital readmission. Although some contain missing values, the amount of missing data is either relatively small or the feature is considered sufficiently informative to justify further preprocessing rather than removal.

Feature: `medical_specialty`

The `medical_specialty` feature records the specialty of the physician responsible for the patient's care. Unlike administrative variables, this feature may contain clinically relevant information because patients treated by different medical specialties may have different risks of hospital readmission. Although approximately **49.08%** of the values are missing, removing the entire feature would result in the loss of potentially valuable information.

Feature: `race`

The `race` feature records the patient's reported racial group and contains only **2.23%** missing values. Since the proportion of missing data is relatively small and the feature may still provide useful demographic information.

Features: `diag_1`, `diag_2` and `diag_3`

These features record the patient's primary, secondary and additional secondary diagnoses respectively. These diagnosis codes provide important clinical information that may help predict hospital readmission.

Only a very small proportion of values are missing (`diag_1`: **0.02%**, `diag_2`: **0.35%**, and `diag_3`: **1.40%**). Removing these features or deleting the affected hospital encounters would result in an unnecessary loss of valuable clinical information.

---


### Features Requiring Special Consideration

In addition to the features discussed above, two laboratory test features require special consideration.
Unlike the previous features, they do not contain missing values. Instead, they contain the valid category **`none`**, which indicates that the corresponding laboratory test was **not performed** rather than representing missing information.

These features are examined separately below to ensure that the category **`none`** is interpreted correctly during preprocessing.

Feature: `max_glu_serum`

The `max_glu_serum` feature records the patient's maximum serum glucose test result. The feature contains four possible categories including **`none`**, where `none` indicates that the laboratory test was not performed rather than representing missing information.

Feature: `A1Cresult`

The `A1Cresult` feature records the patient's HbA1c laboratory test result. Similar to `max_glu_serum`, the category **`none`** indicates that the laboratory test was not performed rather than representing missing information.

Since all categories contain meaningful information, no missing value treatment is required for both features. Theese features will therefore be retained without modification during this preprocessing step.


In [ ]:
# ============================================================
# Remove payer_code
# ============================================================

# Remove the payer_code feature because it contains a large
# proportion of missing values and provides administrative
# rather than clinical information.

df = df.drop(columns=["payer_code"], errors="ignore")

# Display remaining missing values.

print(df.isnull().sum())

# Display the updated dataset dimensions.

print("Updated dataset shape:", df.shape)